File: native/android/app/src/main/java/com/smartmaps/ae/core/LocationFusionEngine.kt
package com.smartmaps.ae.core

import kotlin.math.exp
import kotlin.math.max

/**
 * A/E Location Fusion — GPS + Wi-Fi(network) + cell + BT + IMU gate.
 *
 * Every raw fix from any source lands in [ingest]; [fuse] produces the single
 * authoritative FusedFix per tick as an inverse-variance weighted blend of all
 * sources still inside their freshness window. Sources activate purely by
 * providing data — a provider that never reports simply never contributes
 * (no configuration, no fallback ladder, no "primary" source).
 *
 * The IMU gate: when MovementEngine says STILL with high confidence, position
 * jitter from weak sources is frozen out by heavily weighting the last fused
 * point — GPS wander in a pocket doesn't walk the marker.
 */
class LocationFusionEngine {

    private data class Held(val fix: RawFix)

    private val held = mutableMapOf<FixSource, Held>()
    private var lastFused: FusedFix? = null
    private var smoothedSpeed = 0.0
    private var smoothedHeading = 0.0
    private var headingInit = false

    /** Freshness windows per source, ms. Stale entries stop contributing. */
    private val freshnessMs = mapOf(
        FixSource.GPS to 3000L,
        FixSource.WIFI to 8000L,
        FixSource.CELL to 15000L,
        FixSource.BLUETOOTH to 6000L
    )

    fun ingest(fix: RawFix) {
        if (!fix.point.lat.isFinite() || !fix.point.lng.isFinite()) return
        if (fix.accuracyM <= 0 || !fix.accuracyM.isFinite()) return
        val prev = held[fix.source]
        if (prev != null && fix.timestampMs < prev.fix.timestampMs) return
        held[fix.source] = Held(fix)
    }

    fun current(): FusedFix? = lastFused

    fun reset() {
        held.clear()
        lastFused = null
        smoothedSpeed = 0.0
        headingInit = false
    }

    /**
     * Fuse all fresh sources at [nowMs]. Returns null only before any source
     * has ever reported.
     */
    fun fuse(nowMs: Long, movement: MovementEstimate, dtMs: Long): FusedFix? {
        val fresh = held.entries.filter { (source, h) ->
            nowMs - h.fix.timestampMs <= (freshnessMs[source] ?: 5000L)
        }
        if (fresh.isEmpty()) return lastFused

        // Inverse-variance weights, decayed by age within the window.
        var wSum = 0.0
        var lat = 0.0
        var lng = 0.0
        var accWeighted = 0.0
        val sources = mutableSetOf<FixSource>()
        for ((source, h) in fresh) {
            val ageMs = (nowMs - h.fix.timestampMs).toDouble()
            val window = (freshnessMs[source] ?: 5000L).toDouble()
            val ageFactor = max(0.1, 1.0 - ageMs / window)
            val w = ageFactor / (h.fix.accuracyM * h.fix.accuracyM)
            wSum += w
            lat += h.fix.point.lat * w
            lng += h.fix.point.lng * w
            accWeighted += h.fix.accuracyM * w
            sources += source
        }
        var point = GeoPoint(lat / wSum, lng / wSum)
        var accuracy = accWeighted / wSum

        // IMU stillness gate: confidently STILL → anchor to the last fused
        // point so multi-source jitter can't wander a stationary user.
        val prev = lastFused
        if (prev != null && movement.state == MotionState.STILL && movement.confidence > 0.6) {
            val anchor = 0.85 * movement.confidence
            point = GeoPoint(
                prev.point.lat + (point.lat - prev.point.lat) * (1 - anchor),
                prev.point.lng + (point.lng - prev.point.lng) * (1 - anchor)
            )
        }

        // Speed: best explicit speed among fresh sources (GPS wins by weight),
        // else derived from fused displacement; smoothed exponentially.
        val explicitSpeed = fresh.mapNotNull { it.value.fix.speedMps }.maxOrNull()
        val derivedSpeed = if (prev != null && dtMs > 0) {
            haversineM(prev.point, point) / (dtMs / 1000.0)
        } else null
        val rawSpeed = explicitSpeed ?: derivedSpeed ?: smoothedSpeed
        val k = 1 - exp(-dtMs / 800.0)
        smoothedSpeed += (max(0.0, rawSpeed) - smoothedSpeed) * k

        // Heading: explicit bearing when moving; positional bearing as backup.
        val explicitBearing = fresh.mapNotNull { it.value.fix.bearingDeg }.firstOrNull()
        val positional = if (prev != null && haversineM(prev.point, point) > 1.5) {
            bearingDeg(prev.point, point)
        } else null
        val targetHeading = when {
            explicitBearing != null && smoothedSpeed > 1.0 -> explicitBearing
            positional != null -> positional
            else -> smoothedHeading
        }
        if (!headingInit) {
            smoothedHeading = targetHeading
            headingInit = true
        } else {
            smoothedHeading = (smoothedHeading +
                shortestAngleDeltaDeg(smoothedHeading, targetHeading) * k + 360.0) % 360.0
        }

        val fused = FusedFix(
            point = point,
            accuracyM = accuracy,
            speedMps = smoothedSpeed,
            headingDeg = smoothedHeading,
            sources = sources,
            timestampMs = nowMs
        )
        lastFused = fused
        return fused
    }
}

File: native/ios/SMCore/Core/LocationFusionEngine.swift
// LocationFusionEngine.swift — A/E Location Fusion.
//
// GPS + Wi-Fi + cell + BT + IMU gate: every raw fix from any source lands in
// ingest(); fuse() produces the single authoritative FusedFix per tick as an
// inverse-variance weighted blend of all sources inside their freshness
// window. Sources activate purely by providing data — no configuration, no
// fallback ladder. Semantics mirror the Kotlin engine exactly.

import Foundation

final class LocationFusionEngine {

    private var held: [FixSource: RawFix] = [:]
    private var lastFused: FusedFix?
    private var smoothedSpeed = 0.0
    private var smoothedHeading = 0.0
    private var headingInit = false

    private let freshnessMs: [FixSource: Int64] = [
        .gps: 3000, .wifi: 8000, .cell: 15000, .bluetooth: 6000
    ]

    init() {}

    func ingest(_ fix: RawFix) {
        guard fix.point.lat.isFinite, fix.point.lng.isFinite,
              fix.accuracyM > 0, fix.accuracyM.isFinite else { return }
        if let prev = held[fix.source], fix.timestampMs < prev.timestampMs { return }
        held[fix.source] = fix
    }

    func current() -> FusedFix? { lastFused }

    func reset() {
        held.removeAll()
        lastFused = nil
        smoothedSpeed = 0
        headingInit = false
    }

    func fuse(nowMs: Int64, movement: MovementEstimate, dtMs: Int64) -> FusedFix? {
        let fresh = held.filter { source, fix in
            nowMs - fix.timestampMs <= (freshnessMs[source] ?? 5000)
        }
        if fresh.isEmpty { return lastFused }

        // Inverse-variance weights, decayed by age within the window.
        var wSum = 0.0, lat = 0.0, lng = 0.0, accWeighted = 0.0
        var sources = Set<FixSource>()
        for (source, fix) in fresh {
            let age = Double(nowMs - fix.timestampMs)
            let window = Double(freshnessMs[source] ?? 5000)
            let ageFactor = max(0.1, 1.0 - age / window)
            let w = ageFactor / (fix.accuracyM * fix.accuracyM)
            wSum += w
            lat += fix.point.lat * w
            lng += fix.point.lng * w
            accWeighted += fix.accuracyM * w
            sources.insert(source)
        }
        var point = GeoPoint(lat: lat / wSum, lng: lng / wSum)
        let accuracy = accWeighted / wSum

        // IMU stillness gate: confidently still → anchor to the last fused
        // point so multi-source jitter can't wander a stationary user.
        if let prev = lastFused, movement.state == .still, movement.confidence > 0.6 {
            let anchor = 0.85 * movement.confidence
            point = GeoPoint(
                lat: prev.point.lat + (point.lat - prev.point.lat) * (1 - anchor),
                lng: prev.point.lng + (point.lng - prev.point.lng) * (1 - anchor)
            )
        }

        // Speed: best explicit source speed, else displacement-derived; smoothed.
        let explicitSpeed = fresh.values.compactMap(\.speedMps).max()
        var derivedSpeed: Double?
        if let prev = lastFused, dtMs > 0 {
            derivedSpeed = haversineM(prev.point, point) / (Double(dtMs) / 1000.0)
        }
        let rawSpeed = explicitSpeed ?? derivedSpeed ?? smoothedSpeed
        let k = 1 - exp(-Double(dtMs) / 800.0)
        smoothedSpeed += (max(0, rawSpeed) - smoothedSpeed) * k

        // Heading: explicit bearing when moving; positional bearing as backup.
        let explicitBearing = fresh.values.compactMap(\.bearingDeg).first
        var positional: Double?
        if let prev = lastFused, haversineM(prev.point, point) > 1.5 {
            positional = bearingDeg(prev.point, point)
        }
        let targetHeading: Double
        if let b = explicitBearing, smoothedSpeed > 1.0 {
            targetHeading = b
        } else if let p = positional {
            targetHeading = p
        } else {
            targetHeading = smoothedHeading
        }
        if !headingInit {
            smoothedHeading = targetHeading
            headingInit = true
        } else {
            smoothedHeading = (smoothedHeading +
                shortestAngleDeltaDeg(from: smoothedHeading, to: targetHeading) * k + 360)
                .truncatingRemainder(dividingBy: 360)
        }

        let fused = FusedFix(
            point: point,
            accuracyM: accuracy,
            speedMps: smoothedSpeed,
            headingDeg: smoothedHeading,
            sources: sources,
            timestampMs: nowMs
        )
        lastFused = fused
        return fused
    }
}
